File: android/src/main/kotlin/com/apeirotechnologies/conekta_flutter/ConektaProvider.kt
package com.apeirotechnologies.conekta_flutter

import android.app.Activity
import io.conekta.conektasdk.Card
import io.conekta.conektasdk.Conekta
import io.conekta.conektasdk.Token
import org.json.JSONObject

private const val UNEXPECTED_CARD_TOKEN_ERROR =
    "An unexpected error occurred in the card token process"
private const val UNEXPECTED_CARD_TOKEN_CODE = "unexpected_error"

class ConektaProvider : ConektaInterface {

    private lateinit var activity: Activity
    
    override fun init(activity: Activity) {
        this.activity = activity
    }

    override fun setApiKey(apiKey: String) {
        Conekta.setPublicKey(apiKey)
        Conekta.setApiVersion("0.3.0")
        Conekta.collectDevice(activity)
    }

    override fun getApiKey(): String? =
        if (Conekta.getPublicKey().isNullOrEmpty()) null else Conekta.getPublicKey()

    override fun onCreateCardToken(
        conektaCard: ConektaCard,
        callback: ConektaCardTokenCallback
    ) {
        Token(activity).apply {
            onCreateTokenListener { json -> handleResponse(json, callback) }
            create(
                Card(
                    conektaCard.cardName,
                    conektaCard.cardNumber,
                    conektaCard.cvv,
                    conektaCard.expirationMonth,
                    conektaCard.expirationYear
                )
            )
        }
    }

    private fun handleResponse(json: JSONObject, callback: ConektaCardTokenCallback) {
        try {
            if (json.getString("object") == "error") {
                callback(
                        null,
                        ConektaError.CardTokenError(json.getString("code"), json.getString("message"))
                )
            } else {
                callback(json.getString("id"), null)
            }
        } catch (err: Exception) {
            callback(
                    null,
                    ConektaError.CardTokenError(
                            UNEXPECTED_CARD_TOKEN_CODE,
                            err.message ?: UNEXPECTED_CARD_TOKEN_ERROR
                    )
            )
        }
    }
}

File: ios/Classes/Conekta/ConektaProvider.swift
//
//  ConektaProvider.swift
//  conekta_flutter
//
//  Created by Alfonso Osorio Avilez on 30/08/20.
//

class ConektaProvider: ConektaProtocol {
    private var conekta: Conekta? = nil
    
    func start(view: UIViewController) {
        conekta = Conekta()
        conekta?.viewController = view
        conekta?.collectDevice()
    }
    
    func setApiKey(apiKey: String) {
        conekta?.publicKey = apiKey
    }
    
    func getApiKey() -> String {
        return conekta?.publicKey ?? ""
    }
    
    func onCreateCardToken(conektaCard: ConektaCard, callback: @escaping ConektaCardTokenCallback) {
        let card = Card(number: conektaCard.cardNumber, name: conektaCard.cardName, cvc: conektaCard.cvv, expMonth: conektaCard.expirationMonth, expYear: conektaCard.expirationYear)
        let token = conekta?.token(card: card)
        token?.create(completionHandler: { (response) in
            if let json = response as? Dictionary<String, Any> {
                if json["object"] as? String == "error",
                    let code = json["code"] as? String,
                    let message = json["message"] as? String {
                    callback(nil, CardTokenError(code: code, message: message))
                } else {
                    callback(json["id"] as? String, nil)
                }
            }
        }, errorHandler: { (error) in
            callback(nil, CardTokenError(code: error.localizedDescription, message: error.description))
        })
    }
}
